pkg/chunk/cache/fifo_cache.go

```diff
@@ -166,12 +185,23 @@ func (c *FifoCache) Store(ctx context.Context, keys []string, bufs [][]byte) {
 
 // Stop implements Cache.
 func (c *FifoCache) Stop() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.entries = make(map[string]*cacheEntry)
+	c.first = nil
+	c.last = nil
+	c.currSizeBytes = 0
+
+	c.entriesEvicted.Add(float64(len(c.entries)))
+	c.entriesCurrent.Set(float64(0))
+	c.memoryBytes.Set(float64(0))
 }
 
 // Put stores the value against the key.
 func (c *FifoCache) Put(ctx context.Context, keys []string, values []interface{}) {
 	c.entriesAdded.Inc()
-	if c.size == 0 {
+	if c.maxSizeBytes == 0 && c.maxSizeItems == 0 {
 		return
 	}
```

Comment:
Normal situation is to have one of these non-zero, so check will always fail. I suggest we remove it.

Reply:
Why should we remove it? The check here is used to avoid storing items in the cache if the cache is disabled (which is when the max items/bytes are set to 0).

Reply:
Fifo cache can be disabled via `enable_fifocache` flag at pkg/chunk/cache/cache.go:66. I would suggest that `NewFifoCache` returns nil, if both max size and items are zero, and then cache is not created. Handling this specially in `Put` function is design smell... in my opinion, created cache should never have both parameters set to zero.

Reply:
Design wise I do agree with you. I'm just wondering if what you propose may be considered a breaking change (which we can't afford at this stage).

Reply:
You mean breaking change in terms of API? Are we going to adhere to API versioning rules now?

I believe we consider all Cortex packages as internal packages (with the exception of integration tests framework). (Btw, I checked that Loki doesn't use this)

Functionality-wise, it's not a breaking change, although it may affect some metrics.

Reply:
I did mean breaking change in terms of config, but you're right: there's no config breaking change in your proposal:

> I would suggest that NewFifoCache returns nil, if both max size and items are zero

Your proposal LGTM 👍